docs/proposals/07-shoot-control-plane-migration.md

```diff
@@ -0,0 +1,206 @@
+kind: ShootState
+metadata:
+  name: my-shoot
+  namespace: garden-core
+gardener:
+- name: ca
+  data:
+    ca.crt: ...
+    ca.key: ...
+- name: ssh-keypair
+  data:
+    id_rsa: ...
+- name:
+...
+extensions:
+- kind: Infrastructure
+  state: ... (Terraform state)
+- kind: ControlPlane
+  purpose: normal
+  state: ... (Certificates generated by the extension)
+- kind: Worker
+  state: ... (Machine objects)
+```
+
+### Gardener extensions changes
+
+All extension controllers which support state migration must have a new `status.state` field and support an annotation `core.gardener.cloud/operation=restore` in the respective Custom Resources.
+
+The state field is populated with whatever data the extension controller requires to restore its state on a different seed cluster. As an example the infrastructure resource must save the terraform state.
+
+**Note**: some states can be too large so it makes sense to consider adding them as configmap or secret references.
+
+```
+apiVersion: extensions.gardener.cloud/v1alpha1
+kind: Infrastructure
+metadata:
+  name: infrastructure
+  namespace: shoot--foo--bar
+spec:
+  type: azure
+  region: eu-west-1
+  secretRef:
+    name: cloudprovider
+    namespace: shoot--foo--bar
+  providerConfig:
+    apiVersion: azure.provider.extensions.gardener.cloud/v1alpha1
+    kind: InfrastructureConfig
+    resourceGroup:
+      name: mygroup
+    networks:
+      vnet: # specify either 'name' or 'cidr'
+      # name: my-vnet
+        cidr: 10.250.0.0/16
+      workers: 10.250.0.0/19
+status:
+  state:
+    terraformState: |
+      {
+          "version": 3,
+          "terraform_version": "0.11.14",
+          "serial": 2,
+          "lineage": "3a1e2faa-e7b6-f5f0-5043-368dd8ea6c10",
+          "modules": [
+              {
+              }
+          ]
+          ...
+      }
+```
+
+
+### Shoot reconciliation flow changes
+
+The only data which must be stored in the `ShootState` by the Garden Controller Manager are secrets (e.g ca for the apiserver). Therefore the `botanist.DeploySecrets` step is changed. It is split into two functions which take a list of secrets that have to be generated.
+- `botanist.GenerateSecretState` Generates certificate authorities and other secrets which have to be persisted in the ShootState and must not be regenerated on the New Seed.
+- `botanist.DeploySecrets` Takes secret data from the `ShootState`, generates new ones (e.g. client tls certificates from the saved certificate authorities) and deploys everything in the Shoot's control plane on the New Seed
+
+
+### ShootState synchronization controller
+
+The ShootState synchronization controller is part of the garden controller manager. It syncs the state of extension custom resources from the shoot namespace to the garden cluster and updates the correspodnign `ShootState` resource's `spec.extensions` field. The sync happens every `X` minutes and always runs before migration starts (if possible).
```

Comment:
Whoa, can't we watch instead of this "`X` minutes" approach? What is the load? And what is the load of polling? A watch is much better, I guess, because we get the data almost instantly and we do something only if there is something to do. Polling is usually always a bad thing here.

Without gardenlet: Controller has a watch on the seed's API server and writes to the garden cluster. --> Watch-triggered Pull
With gardenlet: Controller runs in the seed, watches again the seed's API server and writes to the garden cluster. --> Watch-triggered Push

Reply:
Yes watch is much more efficient in this case.

Reply:
Most of the extension controllers are reconciling the resources only in the maintenance time window(when the shoot controller triggers reconciliation), so for the first implementation sync after/during shoot reconciliation should be fine because this is the time when the state can be changed. Later this can and should be improved with watch triggered sync.

Reply:
Hmm, currently the GCM reconciliation flow waits for the result of the reconciliations done by extension controllers (e.g. `botanist.WaitUntilInfrastructureReady`) so this can work.

Reply:
The only problem would be extensions like the worker controller which currently use MCM and therefore the machineobjects (if we have to move them around) can change independent of shoot reconciliation.